### PR TITLE
bigtable: implement managed backup feature

### PR DESCRIPTION
**Pull request by liubonan**

### CODE REVIEW

**Comment by kolea2** [comment]
@tritone do you mind taking a look at this?

**Comment by tritone** [comment]
> @tritone do you mind taking a look at this?

Happy to! @kolea2 @liubonan is there an open issue or doc for this work that I should look at for context?

Also @liubonan , CI is currently failing due to linting-- you just need to run gofmt on your changed files.

File: bigtable/admin.go [hunk 1]
```diff
@@ -1395,3 +1395,228 @@ func UpdateInstanceAndSyncClusters(ctx context.Context, iac *InstanceAdminClient
 
 	return results, nil
 }
+
+// RestoreTable creates a table from a backup. The table will be created in the same cluster as the snapshot.
```
[the remark this answers is not in the document]
Comment (kolea2) [reply bigtable/admin.go:1399]: I think this should be 'backup'?

File: bigtable/admin.go [hunk 2]
```diff
@@ -1395,3 +1395,228 @@ func UpdateInstanceAndSyncClusters(ctx context.Context, iac *InstanceAdminClient
 
 	return results, nil
 }
+
```
[the remark this answers is not in the document]
Comment (liubonan) [reply bigtable/admin.go:1398]: Hi @tritone, I think we will be at GA when this is merged.

File: bigtable/admin.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (liubonan) [reply bigtable/admin.go]: Correct, fix in the updated PR.

File: bigtable/admin.go [hunk 3]
```diff
@@ -1395,3 +1395,228 @@ func UpdateInstanceAndSyncClusters(ctx context.Context, iac *InstanceAdminClient
...
+		StartTime:   startTime,
+		EndTime:     endTime,
+		ExpireTime:  expireTime,
+		State:       backup.State.String(),
+	}, nil
+}
+
+// BackupIterator is an EntryIterator that iterates over log entries.
+type BackupIterator struct {
+	items    []*BackupInfo
```
[the remark this answers is not in the document]
Comment (liubonan) [reply bigtable/admin.go:1527]: Yeah, I agree there were some inconsistency, but the backup implementation was following the table APIs which use TableInfo. Probably we need a separate cleanup PR which get to a consistent naming.

File: bigtable/admin.go [hunk 4]
```diff
@@ -1395,3 +1395,228 @@ func UpdateInstanceAndSyncClusters(ctx context.Context, iac *InstanceAdminClient
...
+	backupPath := clusterPath + "/backups/" + backup
+
+	req := &btapb.DeleteBackupRequest{
+		Name: backupPath,
+	}
+	_, err := ac.tClient.DeleteBackup(ctx, req)
+	return err
+}
+
+// UpdateBackup updates the backup metadata in a cluster.
```
[the remark this answers is not in the document]
Comment (liubonan) [reply bigtable/admin.go:1598]: Yes, the backend only supports updating expireTime.

File: bigtable/admin.go [hunk 5]
```diff
@@ -1395,3 +1395,228 @@ func UpdateInstanceAndSyncClusters(ctx context.Context, iac *InstanceAdminClient
...
+func (it *BackupIterator) Next() (*BackupInfo, error) {
+	if err := it.nextFunc(); err != nil {
+		return nil, err
+	}
+	item := it.items[0]
+	it.items = it.items[1:]
+	return item, nil
+}
+
+// BackupInfo contains backup metadata.
```
[the remark this answers is not in the document]
Comment (liubonan) [reply bigtable/admin.go]: Sure. Added clarification.

File: bigtable/integration_test.go [hunk 6]
```diff
@@ -1865,6 +1865,141 @@ func TestIntegration_InstanceUpdate(t *testing.T) {
...
+		t.Fatalf("IntegrationEnv: %v", err)
+	}
+	defer testEnv.Close()
+
+	if !testEnv.Config().UseProd {
+		t.Skip("emulator doesn't support backups")
+	}
+
+	timeout := 2 * time.Second
+	if testEnv.Config().UseProd {
```
[the remark this answers is not in the document]
Comment (liubonan) [reply bigtable/integration_test.go]: Sure. Removed.

File: bigtable/integration_test.go [hunk 7]
```diff
@@ -1865,6 +1865,141 @@ func TestIntegration_InstanceUpdate(t *testing.T) {
...
+				return nil, err
+			}
+			infos = append(infos, s)
+		}
+		return infos, err
+	}
+
+	// Delete the table at the end of the test. Schedule ahead of time
+	// in case the client fails
+	defer adminClient.DeleteTable(ctx, table)
```
[the remark this answers is not in the document]
Comment (liubonan) [reply bigtable/integration_test.go]: Sure.

File: bigtable/admin.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tritone) [reply bigtable/admin.go:1556]: Ah yeah. Unfortunately we can't make breaking changes like that without releasing a new major version of the library, but it's something to keep in mind. This seems fine to me in any case.

File: bigtable/admin.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tritone) [reply bigtable/admin.go:1627]: I think this should be clarified in the docs, then.

[the remark this answers is not in the document]
Comment (liubonan) [reply bigtable/admin.go]: Added clarification in the comment.

**Pull request merged**